core/cmd/local_client.go

```diff
@@ -240,20 +233,52 @@ func (cli *Client) runNode(c *clipkg.Context) error {
 	}
 
 	lggr.Info("API exposed for user ", user.Email)
+
+	grp, grpCtx := errgroup.WithContext(rootCtx)
+
 	if err = app.Start(); err != nil {
 		return errors.Wrap(err, "error starting app")
 	}
-	defer func() {
-		lggr.ErrorIf(app.Stop(), "Error stopping app")
-		if err = lggr.Sync(); err != nil {
-			log.Fatal(err)
+
+	grp.Go(func() error {
+		<-grpCtx.Done()
+		if err := app.Stop(); err != nil {
+			return errors.Wrap(err, "error stopping app")
 		}
-	}()
+		return nil
+	})
 
 	lggr.Debug("Environment variables\n", config.NewConfigPrinter(cli.Config))
 
 	lggr.Infow(fmt.Sprintf("Chainlink booted in %.2fs", time.Since(static.InitTime).Seconds()), "appID", app.ID())
-	return (cli.Runner.Run(app))
+
+	// After SIGTERM grace period, we must fail fast, but try closing DB
+	go func() {
+		<-rootCtx.Done()
+		<-time.After(cli.Config.ShutdownGracePeriod())
+
+		// this code should not execute if the application finished within the grace period
+		lggr.Criticalf("Shutdown grace period exceeded (%v), closing DB and exiting...", cli.Config.ShutdownGracePeriod())
+		lggr.ErrorIfClosing(db, "failed to close DB")
+		lggr.ErrorIf(lggr.Sync(), "failed to sync Logger")
+
+		os.Exit(-1)
+	}()
+
+	grp.Go(func() error {
+		err := cli.Runner.Run(grpCtx, app)
+		// In tests we have custom runners that stop the app gracefully,
+		// therefore we need to cancel rootCtx when the Runner has quit.
+		// In a real application, this does noop.
+		select {
+		case <-rootCtx.Done():
+		default:
```

Comment:
Won't this always immediately hit the default clause even in a real application?

Reply:
Default is only executed if no other cases are ready.

Reply:
Right, so if rootCtx is not done it just moves ahead immediately. So what's the point of the `<-rootCtx.Done()` clause?

Reply:
The intention was to avoid cancelling the context twice.. but looks like it is Okay to cancel it many times. I will remove this odd `select`.

Reply:
aaaah gotcha I see now